### PR TITLE
Add more standard way to manage databases sessions

### PR DESCRIPTION
**Pull request by sbrunner**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -457,6 +464,49 @@ add_provider(MemoryMapProvider('rss'))
...
+```python
+    env = c2cwsgiutils.setup_process.bootstrap_application_from_options(args)
+    settings = env["registry"].settings
+```
+
+by:
+
+```python
+    loader = pyramid.scripts.common.get_config_loader(options.config_uri)
+    loader.setup_logging(parse_vars(options.config_vars) if options.config_vars else None)
```
Comment (arnaud-morvan) [inline README.md:506]:
```suggestion
    loader.setup_logging(parse_vars(args.config_vars) if args.config_vars else None)
```

File: acceptance_tests/app/c2cwsgiutils_app/scripts.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import argparse
```
Comment (arnaud-morvan) [inline acceptance_tests/app/c2cwsgiutils_app/scripts.py:1]: scripts.py => scripts/get_hello.py ?

File: c2cwsgiutils/db.py [hunk 3]
```diff
@@ -56,6 +63,7 @@ def setup_session(
 
     Returns: The SQLAlchemy session, the R/W engine and the R/O engine
     """
+    warnings.warn("init function is deprecated; use init and request.dbsession instead")
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:66]:
```suggestion
    warnings.warn("setup_session function is deprecated; use init and request.dbsession instead")
```

File: c2cwsgiutils/db.py [hunk 4]
```diff
@@ -112,6 +120,7 @@ def create_session(
 
     Returns: The SQLAlchemy session
     """
+    warnings.warn("init function is deprecated; use init and request.dbsession instead")
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:123]:
```suggestion
    warnings.warn("create_session function is deprecated; use init and request.dbsession instead")
```

File: c2cwsgiutils/db.py [hunk 5]
```diff
@@ -143,6 +152,7 @@ def _add_tween(
     force_master: Optional[Iterable[str]],
     force_slave: Optional[Iterable[str]],
 ) -> None:
+    warnings.warn("init function is deprecated; use init and request.dbsession instead")
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:155]:
```suggestion
    warnings.warn("_add_tween function is deprecated; use init and request.dbsession instead")
```

File: c2cwsgiutils/db.py [hunk 6]
```diff
@@ -183,3 +193,205 @@ def db_chooser_tween(request: pyramid.request.Request) -> Any:
...
+    else:
+        ro_engine = rw_engine = dbengine
+
+    session_factory = SessionFactory(force_master, force_slave, ro_engine, rw_engine)
+    config.registry["dbsession_factory"] = session_factory
+
+    # make request.dbsession available for use in Pyramid
+    def dbsession(request: pyramid.request.Request) -> sqlalchemy.orm.Session:
+        # hook to share the dbsession fixture in testing
+        dbsession = request.environ.get("app.dbsession")
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:390]: Note that you could patch SessionFactory in tests.
It is not common to add hooks in runtime code for tests.

File: c2cwsgiutils/db.py [hunk 7]
```diff
@@ -183,3 +193,205 @@ def db_chooser_tween(request: pyramid.request.Request) -> Any:
...
+        force_master: The method/paths that needs to use the master
+        force_slave: The method/paths that needs to use the slave
+
+    Returns: The SQLAlchemy session
+    """
+    settings = config.get_settings()
+    settings["tm.manager_hook"] = "pyramid_tm.explicit_manager"
+
+    # hook to share the dbengine fixture in testing
+    dbengine = settings.get("dbengine")
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:369]: Note that you could patch get_engine in tests.
It is not common to add hooks in runtime code for tests.

File: c2cwsgiutils/db.py [hunk 8]
```diff
@@ -183,3 +193,205 @@ def db_chooser_tween(request: pyramid.request.Request) -> Any:
...
+    # make request.dbsession available for use in Pyramid
+    def dbsession(request: pyramid.request.Request) -> sqlalchemy.orm.Session:
+        # hook to share the dbsession fixture in testing
+        dbsession = request.environ.get("app.dbsession")
+        if dbsession is None:
+            # request.tm is the transaction manager used by pyramid_tm
+            dbsession = get_tm_session_pyramid(session_factory, request.tm, request=request)
+        return dbsession
+
+    config.add_request_method(dbsession, reify=True)
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:396]: config.add_request_method(dbsession, "dbsession", reify=True)

File: c2cwsgiutils/db.py [hunk 9]
```diff
@@ -183,3 +193,205 @@ def db_chooser_tween(request: pyramid.request.Request) -> Any:
...
+        else:
+            ro_engine = rw_engine
+    else:
+        ro_engine = rw_engine = dbengine
+
+    session_factory = SessionFactory(force_master, force_slave, ro_engine, rw_engine)
+    config.registry["dbsession_factory"] = session_factory
+
+    # make request.dbsession available for use in Pyramid
+    def dbsession(request: pyramid.request.Request) -> sqlalchemy.orm.Session:
```
Comment (arnaud-morvan) [inline c2cwsgiutils/db.py:388]:
```suggestion
    def create_dbsession(request: pyramid.request.Request) -> sqlalchemy.orm.Session:
```

**Review by arnaud-morvan — COMMENTED** [review]
Are you sure this is still compatible with projects with a global DBSession in model ?

File: c2cwsgiutils/db.py (hunk 6 above)
Comment (sbrunner) [reply c2cwsgiutils/db.py:390]: I don't have any opinion about that, but I just get it from here:
https://github.com/Pylons/pyramid-cookiecutter-starter/blob/23d6491fd5c7014990dfbbff45c07dbfdc8b9be1/%7B%7Bcookiecutter.repo_name%7D%7D/%7B%7Bcookiecutter.repo_name%7D%7D/sqlalchemy_models/__init__.py#L119

File: c2cwsgiutils/db.py (hunk 8 above)
Comment (sbrunner) [reply c2cwsgiutils/db.py:396]: Your sure?
https://github.com/Pylons/pyramid-cookiecutter-starter/blob/23d6491fd5c7014990dfbbff45c07dbfdc8b9be1/%7B%7Bcookiecutter.repo_name%7D%7D/%7B%7Bcookiecutter.repo_name%7D%7D/sqlalchemy_models/__init__.py#L127

File: c2cwsgiutils/db.py (hunk 9 above)
Comment (sbrunner) [reply c2cwsgiutils/db.py:388]: I just use the same name as in the starter:
https://github.com/Pylons/pyramid-cookiecutter-starter/blob/23d6491fd5c7014990dfbbff45c07dbfdc8b9be1/%7B%7Bcookiecutter.repo_name%7D%7D/%7B%7Bcookiecutter.repo_name%7D%7D/sqlalchemy_models/__init__.py#L117

**Comment by sbrunner** [comment]
@arnaud-morvan I don't apply your comments about code that comes directly from Pyramid starter, what do you think about that?

File: c2cwsgiutils/db.py (hunk 8 above)
Comment (arnaud-morvan) [reply c2cwsgiutils/db.py:395]: If name is None then it will be computed from the name of the callable

File: c2cwsgiutils/db.py (hunk 6 above)
Comment (arnaud-morvan) [reply c2cwsgiutils/db.py:389]: ok, so let's keep it like that

File: c2cwsgiutils/db.py (hunk 7 above)
Comment (arnaud-morvan) [reply c2cwsgiutils/db.py:368]: ok, so let's keep it like that

File: c2cwsgiutils/db.py (hunk 9 above)
Comment (arnaud-morvan) [reply c2cwsgiutils/db.py:387]: ok, so let's keep it like that

**Review by arnaud-morvan — APPROVED** [review]
Ok, it was just comments, I didn't tested anything, but everything looks good to me.

**Pull request merged**